src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/BulkUpdateService.cs

```diff
@@ -0,0 +1,659 @@
+using Microsoft.Health.Fhir.Core.Features.Resources.Patch;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Messages.Patch;
+using Microsoft.Health.Fhir.Core.Messages.Upsert;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.JobManagement;
+using Microsoft.VisualBasic;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Polly;
+using Polly.Retry;
+using SharpCompress.Common;
+using static Hl7.Fhir.Model.Parameters;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.Health.Fhir.Core.Features.Persistence
+{
+    public class BulkUpdateService : IBulkUpdateService
+    {
+        private readonly IResourceWrapperFactory _resourceWrapperFactory;
+        private readonly Lazy<IConformanceProvider> _conformanceProvider;
+        private readonly IScopeProvider<IFhirDataStore> _fhirDataStoreFactory;
+        private readonly IScopeProvider<ISearchService> _searchServiceFactory;
+        private readonly ResourceIdProvider _resourceIdProvider;
+        private readonly FhirRequestContextAccessor _contextAccessor;
+        private readonly IAuditLogger _auditLogger;
+        private readonly ILogger<BulkUpdateService> _logger;
+
+        internal const string DefaultCallerAgent = "Microsoft.Health.Fhir.Server";
+
+        public BulkUpdateService(
+            IResourceWrapperFactory resourceWrapperFactory,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IScopeProvider<IFhirDataStore> fhirDataStoreFactory,
+            IScopeProvider<ISearchService> searchServiceFactory,
+            ResourceIdProvider resourceIdProvider,
+            FhirRequestContextAccessor contextAccessor,
+            IAuditLogger auditLogger,
+            ILogger<BulkUpdateService> logger)
+        {
+            _resourceWrapperFactory = EnsureArg.IsNotNull(resourceWrapperFactory, nameof(resourceWrapperFactory));
+            _conformanceProvider = EnsureArg.IsNotNull(conformanceProvider, nameof(conformanceProvider));
+            _fhirDataStoreFactory = EnsureArg.IsNotNull(fhirDataStoreFactory, nameof(fhirDataStoreFactory));
+            _searchServiceFactory = EnsureArg.IsNotNull(searchServiceFactory, nameof(searchServiceFactory));
+            _resourceIdProvider = EnsureArg.IsNotNull(resourceIdProvider, nameof(resourceIdProvider));
+            _contextAccessor = EnsureArg.IsNotNull(contextAccessor, nameof(contextAccessor));
+            _auditLogger = EnsureArg.IsNotNull(auditLogger, nameof(auditLogger));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public async Task<BulkUpdateResult> UpdateMultipleAsync(string resourceType, string fhirPatchParameters, int parallelThreads, bool readNextPage, uint maximumNumberOfResourcesPerQuery, bool isIncludesRequest, IReadOnlyList<Tuple<string, string>> conditionalParameters, BundleResourceContext bundleResourceContext, CancellationToken cancellationToken)
```

Comment:
Is the maximumNumberOfResourcesPerQuery parameter used? I see it being passed around to different methods, but I never see it used anywhere.